### PR TITLE
Issue/3196 fix log level usage

### LINKED ISSUE #3196: Fixing LogLevel definition and usage in whole code base
While trying to solve https://github.com/inmanta/inmanta-lsm/issues/450, I made a lot of test cases fail, and it seems this is because we are not using the LogLevel in a consistent way, through the API.

This needs more investigation.

### PR DESCRIPTION
**Pull request by edvgui**
# Description

Fixing LogLevel class definition and usage.

closes #3196 

# Self Check:

Strike through any lines that are not applicable (`~~line~~`) then check the box

- [x] Attached issue to pull request
- [x] Changelog entry
- [x] Type annotations are present
- [x] Code is clear and sufficiently documented
- [x] No (preventable) type errors (check using make mypy or make mypy-diff)
- [x] Sufficient test cases (reproduces the bug/tests the requested feature)
- [x] Correct, in line with design
- [x] End user documentation is included or an issue is created for end-user documentation (add ref to issue here: )

# Reviewer Checklist:

- [ ] Sufficient test cases (reproduces the bug/tests the requested feature)
- [ ] Code is clear and sufficiently documented
- [ ] Correct, in line with design

### CODE REVIEW

File: src/inmanta/const.py [hunk 1]
```diff
@@ -168,11 +168,40 @@ class LogLevel(Enum):
...
+    @property
+    def to_str(self) -> "StrLogLevel":
+        """
+        Convert an integer formated log level into a string formated log level
+        """
+        return StrLogLevel[self.name]
+
+
+@stable_api
+class StrLogLevel(str, Enum):
```
Comment (edvgui) [inline src/inmanta/const.py:180]: The addition of this class might seem strange, as it is not use for now.  This will actually be consumed by `inmanta-lsm`, to fix the openapi doc of some of its routes.

File: src/inmanta/data/model.py [hunk 2]
```diff
@@ -316,6 +319,31 @@ def level_from_enum_attribute(cls, v: object) -> object:
     kwargs: Dict[str, Optional[ArgumentTypes]] = {}
     timestamp: datetime.datetime
 
+    def dict(
```
Comment (edvgui) [inline src/inmanta/data/model.py:322]: This is really ugly, and it could be avoided if the type of the `level` attribute was changed to `StrLogLevel`.  But this would break the api of `resource_deploy_done`.  Any suggestion welcome

File: tests/server/test_purge_on_delete.py [hunk 3]
```diff
@@ -257,7 +262,7 @@ async def test_purge_on_delete_compile_failed(client, server, clienthelper, envi
         version=version,
         resources=[],
         unknowns=[{"parameter": "a", "source": "b"}],
-        version_info={const.EXPORT_META_DATA: {const.META_DATA_COMPILE_STATE: const.Compilestate.failed}},
+        version_info={const.EXPORT_META_DATA: {const.META_DATA_COMPILE_STATE: const.Compilestate.failed.name}},
```
Comment (edvgui) [inline tests/server/test_purge_on_delete.py:265]: Changing the superclass of `Compilestate` from `Enum` to `IntEnum` meant that the "automatic conversion" of any of its enum is done to its integer value, rather than its string name.  What might it break?

Comment (edvgui) [reply tests/server/test_purge_on_delete.py]: The superclass has been changed to `str` and `Enum`, and the values to strings matching the name.  As this is how it was used everywhere in the code.

File: src/inmanta/const.py [hunk 4]
```diff
@@ -168,20 +168,46 @@ class LogLevel(Enum):
...
 
+    @property
+    def to_str(self) -> "StrLogLevel":
+        """
+        Convert an integer formated log level into a string formated log level
+        """
+        return StrLogLevel[self.name]
+
+
+@stable_api
```
Comment (sanderr) [inline src/inmanta/const.py:179]: If this becomes part of the stable API, could you add it to the docs as well? See [here](https://github.com/inmanta/inmanta-core/blame/master/docs/reference/api.rst#L14-L17) (not sure how to link to a rst source outside of blame view). Depending on whether the methods are considered part of the stable API, you could additionally add them for both classes (add their name after `:members:`.

File: src/inmanta/const.py [hunk 5]
```diff
@@ -168,20 +168,46 @@ class LogLevel(Enum):
...
+        Convert a string formated log level into an integer formated log level
+        """
+        return LogLevel[self.name]
+
 
 INMANTA_URN = "urn:inmanta:"
 
 
-class Compilestate(Enum):
+class Compilestate(str, Enum):
```
Comment (sanderr) [inline src/inmanta/const.py:204]: Have you checked that this is not saved to the database? @wouterdb perhaps an annotation for classes that are included in the database could be useful as well, since they need to remain stable too, unless accompanied by a database schema update.

File: src/inmanta/data/model.py (hunk 2 above)
Comment (sanderr) [reply src/inmanta/data/model.py:322]: Could you elaborate on why this is required? I don't immediately what's changed that would require this method override.

**Review by sanderr — CHANGES_REQUESTED** [review]

File: src/inmanta/const.py (hunk 5 above)
Comment (wouterdb) [reply src/inmanta/const.py:204]: As far as I can see, it goes into the database unvalidated, as it comes from the exporter (yikes).
The exporter explicitly uses the string representation.

File: src/inmanta/data/__init__.py [hunk 6]
```diff
@@ -2196,26 +2196,28 @@ def to_dto(self) -> m.CompileRun:
 
 class LogLine(DataDocument):
```
Comment (wouterdb) [inline src/inmanta/data/__init__.py:2197]: This could use a type annotation for the level field.

File: src/inmanta/data/model.py [hunk 7]
```diff
@@ -316,6 +319,31 @@ def level_from_enum_attribute(cls, v: object) -> object:
...
+            include=include,
+            exclude=exclude,
+            by_alias=by_alias,
+            skip_defaults=skip_defaults,
+            exclude_unset=exclude_unset,
+            exclude_defaults=exclude_defaults,
+            exclude_none=exclude_none,
+        )
+
+        # Overwrite the value of the level with its name as the db expects this
```
Comment (wouterdb) [inline src/inmanta/data/model.py:343]: how is this related to the DB? This is the data transfer object?

**Review by wouterdb — CHANGES_REQUESTED** [review]

File: src/inmanta/data/__init__.py (hunk 6 above)
Comment (edvgui) [reply src/inmanta/data/__init__.py:2197]: I don't understand this comment.  There is no `level` field? There is a `level` attribute in the `_data` dict, is it what you are referring to? Or is it the property?  In the former case, how can I add any type annotation to this?  In the later, it is done already no?

File: src/inmanta/data/model.py (hunk 7 above)
Comment (edvgui) [reply src/inmanta/data/model.py:343]: Here: https://github.com/inmanta/inmanta-core/blob/master/src/inmanta/server/services/resourceservice.py#L460
And I wasn't sure if is wasn't being used the same way in other places so I figure I would change it at the root.  I can try changing it only there and see if it works.

File: src/inmanta/data/model.py (hunk 2 above)
Comment (edvgui) [reply src/inmanta/data/model.py:322]: It might not be after all, discussion going on there: https://github.com/inmanta/inmanta-core/pull/3199#discussion_r699166198

File: src/inmanta/data/model.py (hunk 7 above)
Comment (edvgui) [reply src/inmanta/data/model.py]: This change made this test fail, it really want to get the level name when serializing the logline.  Is there a reason for that? https://github.com/inmanta/inmanta-core/blob/master/tests/test_data_model.py#L74
Or can I change the test?

File: src/inmanta/data/__init__.py (hunk 6 above)
Comment (wouterdb) [reply src/inmanta/data/__init__.py:2197]: Ah yes, I'm talking about the`level` in `_data`. I would add comments to document what is in data.

**Review by wouterdb — APPROVED** [review]

**Review by wouterdb — CHANGES_REQUESTED** [review]
The LogLine DTO is now broken, ....

File: src/inmanta/data/__init__.py (hunk 6 above)
Comment (edvgui) [reply src/inmanta/data/__init__.py:2197]: Done in https://github.com/inmanta/inmanta-core/pull/3199/commits/ffb53a8cfda2a0bc7f1b17dc49162e3d69f918af

Comment (wouterdb) [reply src/inmanta/data/__init__.py:2197]: nice!

**Review by wouterdb — APPROVED** [review]
This is both elegant and scary

File: src/inmanta/const.py [hunk 8]
```diff
@@ -155,33 +155,65 @@ def get_agent_trigger_method(cls, is_full_deploy):
...
+    LogLevel.TRACE: 3,
+    LogLevel.NOTSET: 1,
+}
+
+
+# The following code registers the integer log levels as values
+# in the LogLevel enum.  It allows to pass them in the constructor
+# as if it was an integer enum: LogLevel(50) == LogLevel.CRITICAL
+for level, value in LOG_LEVEL_AS_INTEGER.items():
+    LogLevel._value2member_map_[value] = level
```
Comment (sanderr) [inline src/inmanta/const.py:204]: I agree with @wouterdb's overall assessment. I'm a bit scared of using this private member, but as long as there are tests to guard against this failing I guess it's all right. It is indeed elegant.

**Review by sanderr — APPROVED** [review]

**Review by wouterdb — APPROVED** [review]
